control/server.py

```diff
@@ -30,17 +29,22 @@
 from .state import GatewayState, LocalGatewayState, OmapGatewayState, GatewayStateHandler
 from .grpc import GatewayService
 
-libc = ctypes.CDLL(ctypes.util.find_library("c"))
-PR_SET_PDEATHSIG = 1
-
+def sigchld_handler(signum, frame):
+    """Handle SIGCHLD, runs when a spdk process terminates."""
+    logger = logging.getLogger(__name__)
+    stack = '\n'.join(traceback.format_stack(frame))
+    logger.error(f"GatewayServer: GSIGCHLD received {signum=} stack=\n{stack}")
 
-def set_pdeathsig(sig=signal.SIGTERM):
+    try:
+        pid, wait_status = os.waitpid(-1, os.WNOHANG)
+    except OSError:
+        logger.exception(f"waitpid error:")
+        # eat the exception, in signal handler context
 
-    def callable():
-        return libc.prctl(PR_SET_PDEATHSIG, sig)
-
-    return callable
+    exit_code = os.waitstatus_to_exitcode(wait_status)
 
+    # GW process should exit now
+    raise SystemExit(f"spdk subprocess terminated pid {pid=} {exit_code=}")
```

Comment:
"pid" word is not needed here, because we have with it "spdk subprocess terminated pid pid=..."